qkeras/quantizers.py

```diff
@@ -439,9 +443,517 @@ def trainable_variables(self):
+          multiplied by the data type scale.
+      keep_negative (bool): If false, we clip negative numbers. Defaults to 
+        True.
+      use_stochastic_rounding (bool): If true, we perform stochastic rounding
+        (https://arxiv.org/pdf/1502.02551.pdf).
+      scale_axis (int, None): Which axis to calculate scale from. If None, we
+        perform per-channel scaling based off of the image data format. Note 
+        that each entry of a rank-1 tensor is considered its own channel by
+        default. See `_get_scaling_axis` for more details. Defaults to None.
+      qnoise_factor (float): A scalar from 0 to 1 that represents the level of
+        quantization noise to add. This controls the amount of the
+        quantization noise to add to the outputs by changing the weighted
+        sum of (1 - qnoise_factor) * unquantized_x + qnoise_factor *
+        quantized_x. Defaults to 1.0, which means that the result is fully
+        quantized.
+      use_variables (bool): If true, we use tf.Variables to store certain 
+        parameters. See the BaseQuantizer implementation for more details.
+        Defaults to False. If set to True, be sure to use the special attribute
+        update methods detailed in the BaseQuantizer.
+      var_name (str or None): A variable name shared between the tf.Variables
+        created in on initialization, if use_variables is true. If None, the 
+        variable names are generated automatically based on the parameter names 
+        along with a uid. Defaults to None.
+
+    Returns:
+      function: Function that computes linear quantization.
+
+    Raises:
+      ValueError:
+        - If `bits` is not positive, or is too small to represent `integer`.
+        - If `integer` is negative.
+        - If `alpha` is a string but not one of ("auto", "auto_po2").
+
+    """
+
+  # string options for alpha parameter
+  ALPHA_OPTIONS = ("auto", "auto_po2")
+
+  def __init__(
+      self,
+      bits=8,
+      integer=0,
+      symmetric=1,
+      keep_negative=True,
+      alpha=None,
+      use_stochastic_rounding=False,
+      scale_axis=None,
+      qnoise_factor=1.0,
+      var_name=None,
+      use_variables=False,
+  ):
+    super(quantized_linear, self).__init__()
+
+    # Set _initialized parameter to False to prevent the setters from
+    # performing preliminary calculations
+    self._initialized = False
+    self.var_name = var_name
+    self.bits = bits
+    self.integer = integer
+    self.symmetric = symmetric
+    self.keep_negative = keep_negative
+    self.qnoise_factor = qnoise_factor
+    self.use_stochastic_rounding = use_stochastic_rounding
+    self.alpha = alpha
+    self.scale_axis = scale_axis
+    self.use_variables = use_variables
+
+    # Perform preliminary calculations based on attributes above
+    self._initialized = True
+    self._set_default_quantization_scale()
+
+
+  @property
+  def bits(self):
+    return self._bits
+
+  @bits.setter
```

Comment:
I think that makes sense. If the intention is to not have the attributes be modifiable, then we ought to enforce that more strictly. As a small example, this would look like

```
def __init__(self, bits):
  self._check_bits(bits)
  self._bits = bits
  
def _check_bits(self, bits):
  err_msg = f"Bit count {bits} must be positive"
  if bits <= 0:
    raise ValueError(err_msg)

@property
def bits(self):
  return self._bits
```

This way the user would receive an error when trying to run
```
q = quantized_linear(bits=8)
q.bits = 2
```

The option of using NO `@property` decorators is not preferable in my mind for attributes that need validation checks

Reply:
If we make these parameters as trainable weights, this will require too great of a refactoring for all of QKeras as previously discussed - so I don't think that is a feasible option. The other solution of removing the option to change these parameters will limit us unnecessarily in case we want to put these parameters on a schedule with a Keras callback when training. Therefore, I think the current solution is best.

Reply:
After discussing with Shan, there is risk in changing the bits and other parameters during training since it could collapse or make the training unstable. So for now let's remove these properties and re-implement it properly later on when we have more information.

Reply:
Okay, that's fine. In that case I will get rid of these lines:

https://github.com/google/qkeras/blob/master/qkeras/quantizers.py#L399-L405

I doubt that this was even properly implemented since there is no update method (like there is for qnoise_factor)

Reply:
I've made the requested change. I've included some extra checks to ensure that non-modifiable attributes are non-modifiable. See https://www.pythonmorsels.com/making-read-only-attribute/